packages/medusa/src/models/batch-job.ts

```diff
@@ -1,30 +1,91 @@
-import { BeforeInsert, Column, Entity, PrimaryColumn } from "typeorm"
+import { AfterLoad, BeforeInsert, Column, Entity, JoinColumn, ManyToOne } from "typeorm"
 import { BatchJobStatus } from "../types/batch-job"
-import { DbAwareColumn } from "../utils/db-aware-column"
+import { DbAwareColumn, resolveDbType } from "../utils/db-aware-column"
 import { SoftDeletableEntity } from "../interfaces/models/soft-deletable-entity"
 import { generateEntityId } from "../utils/generate-entity-id"
+import { User } from "./user"
 
 @Entity()
 export class BatchJob extends SoftDeletableEntity {
   @DbAwareColumn({ type: "text" })
   type: string
 
-  @DbAwareColumn({ type: "enum", enum: BatchJobStatus })
-  status: BatchJobStatus
-
-  @Column({ type: "text", nullable: true })
+  @Column({ nullable: true })
   created_by: string | null
 
+  @ManyToOne(() => User)
+  @JoinColumn({ name: "created_by" })
+  created_by_user: User
+
   @DbAwareColumn({ type: "jsonb", nullable: true })
-  context: Record<string, unknown>
+  context: { retry_count?: number; max_retry?: number } & Record<string, unknown>
 
   @DbAwareColumn({ type: "jsonb", nullable: true })
   result: Record<string, unknown>
 
+  @Column({ type: "boolean", nullable: false, default: false })
+  dry_run: boolean = false;
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  pre_processing_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  processing_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  confirmed_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  completed_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  canceled_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  ready_at?: Date
```

Comment:
cc: @olivermrbl @pKorsholm

Reply:
Do you mean removing ready and instead combined pre_processing and ready to pre_processed?

the flow would be then
created -> pre_processed -> confirmed (if dry run) -> processing -> complete?

Reply:
we could use preprocessing_at, it depends on how you want to signify that the worker can take over and begin to process the batch job. The essential part is that we can support the following 3 phases of processing a batch job with a possibility to halt execution between preprocessing and processing, below described in terms of an import with dry_run enabled:
1. setup (done by the API)
remote files are uploaded and invariants for batch job are established

(at this point Ready was set)

2. preprocessing (done by the worker) 
Parsing and preparation before anything is written to the database

(at this point we wait for confirmation if dry_run is enabled)

3. processing (done by the worker)
Writing to the database

Ready_at and Preprocessing_at seems like two views on a state from the point of view of the api and the worker respectively. There is a possibility of a period between the states, where the batchjob is ready but not preprocessing. If we want encapsulate the waiting time as well we should introduce more states instead of trying to reduce the number of states. 
The same is applicable when moving from "confirmed" into "processing" which is the working phase that initiated this discussion.

Reply:
I think that seb proposal is providing the same aspects we've discussed but with one less status.

Basically here is what I've understood about the proposal flow:
- CREATED -> The job has been created and the worker can pick it up as soon as possible
- Now the worker is pre processing the job and one it is done it will marked it as PRE-PROCESSED
- PRE_PROCESSED, we are now in the waiting step until confirmation if it is a dry run job or the pre processed will marked it immediately  as confirmed.
The rest after confirmation is the same.

Reply:
I dont like using the CREATED to communicate to the worker to pick up a batch job. Mostly because any pre-processing such as file uploads, or whatever people want to do prior to creating the batch job would happen before it is created in the database. This is why we had the READY status previously and why I think we should keep it.

Reply:
After discussion here is the output is that
- CREATED the job entity had been created been created, but at the same time it might be still doing some stuff like sending a file or something so we should have a ready_to_be_preprocessed intermediary status
- from that ready to be pre processed the worker can pick up the job for pre processing
- Here the pre processing is done and the job is ready
- Then either the core confirm it or the user depending on dry run (as it is now)
- Then after the confirmation the job can be picked up by the worker to be process
- And then complete the job

Reply:
@srindom After our correspondance we have changed it back to use the `created` state as a trigger for the worker. The states at this time are: 

BatchJobStatus.CREATED -> triggers worker, at this point the batch job is prepared 
BatchJobStatus.PRE_PROCESSED -> The worker has finished preprocessing, and for `dry_run` it pauses
BatchJobStatus.CONFIRMED -> The worker can pick up the batch job again for processing
BatchJobStatus.PROCESSING -> the worker has picked up the batch job for processing
BatchJobStatus.COMPLETED -> Processing is complete

there are also FAILED and CANCELLED which remain unchanged. @adrien2p can you confirm this is correct?

Reply:
yes this seams to be correct 👍